src/js/controllers/mapController.ts

```diff
@@ -402,6 +417,227 @@ export class MapController {
     this._sketchVM?.create('polygon', { mode: 'freehand' });
   };
 
+  getAndDispatchMeasureResults(optionType: string): void {
+    let areaResults = {};
+    let distanceResults = {};
+    this._selectedWidget?.watch(
+      'viewModel.measurementLabel',
+      (
+        measurementLabel: AreaMeasurement2D['viewModel']['measurementLabel']
+      ) => {
+        switch (optionType) {
+          case 'area': {
+            if (measurementLabel) {
+              areaResults = {
+                area: measurementLabel.area,
+                perimeter: measurementLabel.perimeter
+              };
+            }
+            break;
+          }
+          case 'distance': {
+            if (measurementLabel) {
+              distanceResults = {
+                length: measurementLabel
+              };
+            }
+            break;
+          }
+          case 'coordinates':
+            // do something
+            break;
+          default:
+            break;
+        }
+      }
+    );
+    this._selectedWidget?.watch('viewModel.state', (state: string) => {
+      if (state === 'measured') {
+        store.dispatch(
+          setMeasureResults({
+            areaResults,
+            distanceResults,
+            coordinateMouseClickResults: {},
+            coordinatePointerMoveResults: {}
+          })
+        );
+      }
```

Comment:
we can extract area/perimeter/distance from selectedWidget in here, no need for the above watcher!

Reply:
this PR